File: Frontend/apps/astrogame_app/lib/models/buildings/building.dart
import 'package:astrogame_app/communications/converters/building_converter.dart';
import 'package:astrogame_app/communications/converters/guid_converter.dart';
import 'package:astrogame_app/models/enums/building_type.dart';
import 'package:astrogame_app/models/enums/stellar_object_type.dart';
import 'package:astrogame_app/models/technologies/technology.dart';
import 'package:json_annotation/json_annotation.dart';

import 'input_resource.dart';
import 'output_resource.dart';

@GuidConverter()
abstract class Building extends Technology {
  @JsonKey()
  late BuildingType buildingType;

  @JsonKey()
  late StellarObjectType buildableOn;

  @JsonKey()
  List<InputResource>? inputResources = [];

  @JsonKey()
  List<OutputResource>? outputResource = [];

  Building();

  factory Building.fromJson(Map<String, dynamic> json) =>
      BuildingConverter().fromJson(json);
}

File: Backend/AstroGame.Shared/Models/Buildings/Building.cs
﻿using AstroGame.Shared.Enums;
using System.Collections.Generic;
using AstroGame.Shared.Models.Technologies;

namespace AstroGame.Shared.Models.Buildings
{
    public abstract class Building : Technology
    {
        public BuildingType BuildingType { get; set; }
        
        /// <summary>
        /// Sets the StellarObject types where this building can be built on
        /// </summary>
        public StellarObjectType BuildableOn { get; set; } = new();

        //public virtual List<BuiltBuilding> BuiltBuildings { get; set; } = new();
        public List<InputResource> InputResources { get; set; } = new();
        public List<OutputResource> OutputResources { get; set; } = new();
    }
}